insights/combiners/satellite_version.py

```diff
@@ -101,43 +42,115 @@ def _parse_sat_versoin(version):
+                   the version, here checked `satellite-schema` instead of
+                   `satellite-branding`.
+
+           Check the version of package satellite-schema directly:
+
+                                        Sat 5.0~5.2     Sat 5.3 ~
+                rhn-satellite-schema    ok              -
+                satellite-schema        -               ok
+
+    Attributes:
+        full(str): the full version format like `version-release`.
+        version(str): the satellite version do not includes `release`.
+        release(str): the `release` string in the version.
+        major(int): the major version.
+        minor(int): the minor version.
+
+    Examples:
+        >>> sat_ver = shared[satellite_version]
+        >>> sat_ver.major
+        6
+        >>> sat_ver.minor
+        2
+        >>> sat_ver.version
+        '6.2.2'
+        >>> sat_ver.full
+        '6.2.2.1-1.0.el7sat'
+        >>> sat_ver.release
+        '1.0.el7sat'
+
     """
-    # For Satellite 6.1.x, if satellite_version/version.rb is available:
-    if sat6_ver and sat6_ver.version:
-        return SatelliteVersion(sat6_ver.full, sat6_ver.version, None,
-                                sat6_ver.major, sat6_ver.minor)
-
-    if rpms:
-        # For Satellite 6.2.x, check the ``satellite`` package directly
-        sat62_pkg = rpms.get_max('satellite')
-        if sat62_pkg:
-            version = sat62_pkg.version
-            major, minor = _parse_sat_versoin(version)
-            return SatelliteVersion(sat62_pkg.package, version,
-                                    sat62_pkg.release, major, minor)
-
-        # For Satellite 6.0.x/6.1.x, check the version of:
-        # - foreman, candlepin and katello
-        fman = rpms.get_max('foreman')
-        cndp = rpms.get_max('candlepin')
-        ktlo = rpms.get_max('katello')
-        if fman and cndp and ktlo:
-            for sat_ver, map_ver in sat6_ver_map.items():
-                if all(pkg.version.startswith(mv)
-                        for pkg, mv in zip([fman, cndp, ktlo], map_ver)):
-                    major, minor = _parse_sat_versoin(sat_ver)
-                    return SatelliteVersion(sat_ver, sat_ver, None, major, minor)
-
-        # For Satellite 5.x
-        sat5_pkg = rpms.get_max('satellite-schema')
-        sat5_pkg = sat5_pkg if sat5_pkg else rpms.get_max('rhn-satellite-schema')
-        if sat5_pkg:
-            version = sat5_pkg.version
-            major, minor = _parse_sat_versoin(version)
-            return SatelliteVersion(sat5_pkg.package, version,
-                                    sat5_pkg.release, major, minor)
+    def __init__(self, sat6_ver, rpms):
+        self.full = None
+        self.version = None
+        self.release = None
+        self.major = None
+        self.minor = None
+
+        # For Satellite 6.1.x, if satellite_version/version.rb is available:
+        if sat6_ver and sat6_ver.version:
```

Comment:
If this condition and the ones below it within "if rpms:" are all false (i.e., this isn't a satellite machine), then the combiner will not be correctly initialized in almost every case since InstalledRpms will almost always be available to satisfy one of its dependencies. If the combiner doesn't initialize correctly, it should raise `SkipComponent` (from `insights.core.dr`), or we should move `SkipException` up out of `insights/parsers/__init__.py` to a higher level and have it raise that.

Reply:
In fact, I have once tried to raise an `Exception` for the case it's not a satellite machine (just like what 1.x did). However, just like you said, the `InstalledRpms` will almost always be available. Then for most situations which are not satellite machine, the `Exception` will be raised (it seems like something is wrong, but it's just not a satellite machine). So I left (initialized) all attributes as `None` when it's not a satellite machine instead of raising an `Exception`.  (And I'm afraid that these `Exception` might be treated as a kind of `Parser Error` and will cause maintainer's unnecessary attention). Does that make sense to you?

Reply:
Hey @xiangce, `insights.core.dr.SkipComponent` is the general `Exception` class in master that should be raise if a component wants to take itself out of dependency resolution and not have it counted as something going wrong. If a component raises `SkipComponent` or a subclass of it (like `insights.parsers.SkipException in master), the exception is ignored.

I think we might need to create something like `SkipComponent` in the 1.x branch and then have the `satellite_version` combiner raise it in both master and 1.x. I think proper handling of `insights.parsers.SkipException` needs to be added to 1.x anyway.

Let's hold this PR for now, and I'll chat with Kyle about some changes to 1.x core that it can use.

Reply:
@csams - Thanks for your explanation, that sounds good. I will modify this PR at first.